index.bs

```diff
@@ -2801,13 +2945,29 @@ The [=remote end steps=] with |command parameters| are:
   1. If the implementation is unable to create a new browsing context for any
      reason then return [=error=] with [=error code=] [=unsupported operation=].
 
+  1. Let |user context| be the [=default user context=] if |reference context| is null, and |reference context|' [=user context=] otherwise.
+
+  1. If |command parameters|[<code>userContext</code>] is present and is not null:
+
+    1. Let |user context id| be |command parameters|[<code>userContext</code>].
+
+    1. If the [=user context=] with the [=user context id=] equal to |user context id| does
+      not exist in the [=remote end=], return [=error=] with [=error code=] [=no such user context=].
+
+    1. Set |user context| to the [=user context=] with the [=user context id=] equal
+      to |user context id|.
+
+    1. If |reference context| is not null and |reference context|'s [=user context=] is not |user context|,
+      return [=error=] with [=error code=] [=invalid user context id=].
```

Comment:
In Firefox, we can open have tabs from several user contexts (containers) next to each other in the same window. I imagine it's fine to enforce this limitation if other vendors don't have this flexibility. 

Does that mean that in Chrome, if you have 1 window with 1 tab in the default context and you use `browsingContext.create` with type=tab, it will force creating a new window on the fly? In that case it's not compliant with the following step of the type=tab section:

> The new browsing context should reuse an existing OS window, if any.

So I'm wondering if we really need to be strict and fail here. We could modify the text of the "if type is tab" section below to say we will "attempt" to reuse an existing window (containing the reference context, if relevant), but it's not guaranteed.

cc @jgraham

Reply:
> In Firefox, we can open have tabs from several user contexts (containers) next to each other in the same window. I imagine it's fine to enforce this limitation if other vendors don't have this flexibility.

Yes, that is not supported with Chrome.

> Does that mean that in Chrome, if you have 1 window with 1 tab in the default context and you use browsingContext.create with type=tab, it will force creating a new window on the fly? In that case it's not compliant with the following step of the type=tab section:

> The new browsing context should reuse an existing OS window, if any.

Yes, it will create a tab in a new window. I think "should" allows for that.

Reply:
Ah maybe, I thought the "should" was just there to cover the case where there is no existing window ("if any"). But maybe this also covers the cases where existing windows are not compatible with the tab we are trying to add.

The spec about the reference context uses the same phrasing with "should":

> If reference context is not null, the new browsing context should reuse the window containing reference context, if any.

So if the "window containing the reference context" is not compatible with the tab we are trying to add, then it should be fine to add the tab in another window? 

Again, not sure we should throw here, if the spec is flexible enough to allow to pick another window if needed?

Since nothing in Firefox prevents from having tabs from several containers in the same window, it seems a bit arbitrary to throw. Imagine we have 1 window with 1 tab "tabA" owned by "contextA". If someone uses "browsingContext.create(type=tab, userContext=contextB)" with no reference context, Firefox will add this tab next to tabA. But if the user passes referenceContext=tabA, we would throw because contexts don't match?

If we really want to have a consistent behavior across all browsers, I think we need to enforce that all tabs in a window all have the same user context, but that seems difficult/impossible to implement.

Reply:
"should" in specs is defined as "unless there's a valid reason not to". To me the "should reuse the window containing the reference context" is fine; it more or less macro expands to "shall reuse the window containing the reference context, unless there's a valid reason not to", and "we don't support multiple user contexts in the same window" seems like potentially a valid reason.

I also agree that Firefox in general can't enforce the condition that each OS window only gets tabs with a single user context and therefore we can't in WebDriver. For example we might connect to a browser where the user already created multiple containers in the same window.

Given that implementation differences seem impossible to avoid here, I think it might be OK to just allow whatever behaviour the implementation does? In practice clients can always request a new window when creating a tab in a different context, so for test automation purposes it's possible to build a stable, interoperable, foundation despite the underlying differences.